platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/roster/RosterDiffGenerator.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.platform.roster;
+
+import static com.swirlds.logging.legacy.LogMarker.EXCEPTION;
+
+import com.swirlds.common.context.PlatformContext;
+import com.swirlds.common.crypto.Cryptography;
+import com.swirlds.common.system.NodeId;
+import com.swirlds.common.system.address.Address;
+import com.swirlds.common.system.address.AddressBook;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+// TODO unit test
+
+/**
+ * Takes as input a sequence of {@link UpdatedRoster} objects and produces a sequence of {@link RosterDiff} objects.
+ */
+public class RosterDiffGenerator {
+
+    private static final Logger logger = LogManager.getLogger();
+
+    private AddressBook previousRoster;
+    private long previousEffectiveRound;
+
+    private final Cryptography cryptography;
+
+    /**
+     * Constructor.
+     *
+     * @param platformContext the platform context
+     */
+    public RosterDiffGenerator(@NonNull final PlatformContext platformContext) {
+        cryptography = platformContext.getCryptography();
+    }
+
+    /**
+     * Given a new roster, generate a diff with respect to the previous roster.
+     *
+     * @param updatedRoster the new roster
+     * @return the diff, will return null for the very first roster added
+     */
+    @Nullable
+    public RosterDiff generateDiff(@NonNull final UpdatedRoster updatedRoster) {
+        final AddressBook roster = updatedRoster.roster();
+        final long effectiveRound = updatedRoster.effectiveRound();
+
+        try {
+            if (previousRoster == null) {
+                // This is the first roster we've seen.
+                return null;
```

Comment:
When I see a return, I want to know that the private data fields are determined.  Wrapping return statements in `try-finally` blocks means I have to know the impacts of everything in the finally block and it is easy to miss they exist.  (A linear reading of the code is not possible, I have to know ahead of time that I need to jump to the finally block to get the complete picture.)    Personally I think this pattern is a code-smell.   It saves lines of code, but makes the code more complicated to review.

Reply:
In this case I was going to observe that the `previousRoster` and `previousEffectiveRound` were not set before returning.   But then I realized the finally was doing that work.

Reply:
refactored to be linear

Reply:
much appreciated